src/exoticatechnologies/hullmods/exotics/HullmodExoticHandler.kt

```diff
@@ -18,183 +31,851 @@ object HullmodExoticHandler {
+            // But for LENIENT - lets just check if it's in the variants list and then improve from there ...
+            // On the other hand/thought - why not try making these two the same?
+            // UPDATE: for some reason, the list of variants we get from the root module *does not* equal any of the variants
+            // in the "all module variants" list, so similar to before with the 'is expected' - lets check via hullspec hullID
+
+            //TODO update documentation so that it refers to both LENIENT and STRICT
+            if (workMode == HullmodExoticHandlerWorkMode.STRICT) {
+                // STRICT work mode - validate if the variant is both in the 'installed on' list and 'module variants' list
+                val isInstalledOn = currentInstallData.listOfVariantsWeInstalledOn.contains(variant)
+                val isInVariantsList = currentInstallData.listOfAllModuleVariants.contains(variant)
+                // And just return that, since we should allow uninstalling if we have it installed or prevent uninstalling if we dont
+                return isInstalledOn and isInVariantsList
+            } else {
+                // LENIENT work mode - similar to strict, except we won't be doing a literal contains() but rather a slightly
+                // more lax check - via the hullSpec.hullID
+                val isInstalledOn = currentInstallData.listOfVariantsWeInstalledOn.contains(variant)
+                val isInVariantsList = currentInstallData.listOfAllModuleVariants
+                        .map { moduleVariant -> moduleVariant.hullSpec.hullId }
+                        .contains(variant.hullSpec.hullId)
+                // And just return whether we're both installed and in the variants list
+                return isInstalledOn and isInVariantsList
+            }
+        }
     }
 
-    fun removeHullmodExoticFromFleetMember(exoticHullmod: ExoticHullmod, fleetMember: FleetMemberAPI) {
-        val keysToRemove = mutableSetOf<HullmodExoticKey>()
-        val allKeys = grabAllKeysForParticularFleetMember(fleetMember)
-        for (key in allKeys) {
-            val exoticHandlerDataOptional = getDataForKey(key)
-            if (exoticHandlerDataOptional.isPresent()) {
-                val exoticHandlerData = exoticHandlerDataOptional.get()
-                for (variant in exoticHandlerData.listOfVariantsWeInstalledOn) {
-                    val variantHullSize = variant.hullSpec.hullSize
-                    exoticHullmod.removeEffectsBeforeShipCreation(variantHullSize, variant.statsForOpCosts, exoticHullmod.hullModId)
-                    // Lets keep track of 'removed' keys here because why not
-                    keysToRemove.add(key)
+    private fun removeHullmodExoticFromVariant(hullmodExotic: HullmodExotic, parentFleetMember: FleetMemberAPI, variant: ShipVariantAPI, workMode: HullmodExoticHandlerWorkMode): Boolean {
```

Comment:
javadoc please. I won't remember what's the whole point of this method later on...

Reply:
done. resolving.